Sources/SwiftProtobuf/Data+Extensions.swift

```diff
@@ -0,0 +1,49 @@
+// Sources/SwiftProtobuf/Data+Extensions.swift - Extension wrapping deprecated Data methods
```

Comment:
Elsewhere, we've done this the other way around.

That is, update the code to use the _new_ syntax, and write an adapter to simulate the new syntax on old Swift versions.  It's a bit more work, but means that when we stop supporting those old versions, we can simply delete the adapter code.

Reply:
If you could get the adapter right (simulating new syntax on old Swift), and prove that it works by converting just a few places, I'm happy to help get the rest converted.  I've gotten pretty handy at scripting global changes of this sort.

Reply:
So I took a look at it, and I think doing it in this fashion is maybe a bit worse than the current one. 

My opinion is based on this [tweet](https://twitter.com/jckarter/status/1071183859083960320) — basically Apple is saying that we should carefully check that the pointer is both valid and has a length (`count`).

For example, if we try to support the new syntax in the code, this call site (from `BinaryDelimited.swift`): 

```
    data.withUnsafeBytes { (pointer: UnsafePointer<UInt8>) in
      written = stream.write(pointer, maxLength: totalSize)
    }
```

Would have to be converted to this:

```
    data.withUnsafeBytes { (body: UnsafeRawBufferPointer) in
      if let baseAddress = body.baseAddress, body.count > 0 {
        let pointer = baseAddress.assumingMemoryBound(to: UInt8.self)
        written = stream.write(pointer, maxLength: totalSize)
      }
    }
```

But if someone tries to write this again, they might use the old way and forget to do the `count` check and the `baseAddress` check, leading to a crash. With the wrapper, we can make sure that no future additions will crash, and when we want to delete the old code we just remove the `#else` case in the wrapper.

In addition, if we want to convert all the call sites to the new API, we're going to have a problem because the new API in 5.0 is this:

```
func withUnsafeBytes<ResultType>(_ body: (UnsafeRawBufferPointer) throws -> ResultType) rethrows -> ResultType
```

The new function no longer is generic over the content type in the Data like the old API:

```
withUnsafeBytes<ResultType, ContentType>(_ body: (UnsafePointer<ContentType>) throws -> ResultType) rethrows -> ResultType
```

But in the implementation of the closure, you need to know the content type so you can call `assumingMemoryBound(to:)`:

```
let pointer = baseAddress.assumingMemoryBound(to: UInt8.self)
```

I can't find a way to provide that type to `assumingMemoryBound(to:)` at runtime, it needs to be provided at compile time. Either we use the old API approach of genericizing over `ResultType` and `ContentType` (which defeats the purpose of making this change), or we take in the type in the function:

```
internal func withUnsafeBytes<T, ResultType>(innerType: T.Type, _ body: (UnsafeRawBufferPointer) throws -> ResultType) rethrows -> ResultType {
    if count > 0 {
      return try withUnsafeBytes { (pointer: UnsafePointer<T>) -> ResultType in
        return try body(UnsafeRawBufferPointer(start: pointer, count: count))
      }
    }
  }
```

But then we're still not providing the exact same new API to older versions of Swift. So perhaps my approach is the only one that works for now, even though it will require changing the call sites when we remove `>=4.1` support.

But I'll defer to your preference on this, happy to do either approach.